File: problems/basic/23/solution1.py
# Accepted: 71ms (96.92%), 20.80MB (16.92%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        queue = []

        for nodeList in lists:
            node = nodeList
            while node is not None:
                heappush(queue, node.val)
                node = node.next

        cur = mergedHead = ListNode(0)
        while len(queue) > 0:
            val = heappop(queue)
            cur.next = ListNode(val)
            cur = cur.next

        return mergedHead.next

File: problems/basic/23/solution1.cpp
// Accepted: 11ms (92.74%), 17.67MB (17.27%)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<int, vector<int>, greater<int>> queue;

        for (ListNode* node : lists) {
            while (node != nullptr) {
                queue.push(node->val);
                node = node->next;
            }
        }

        ListNode mergedHead;
        ListNode *cur = &mergedHead;
        while (!queue.empty()) {
            int val = queue.top();
            queue.pop();
            cur->next = new ListNode(val);
            cur = cur->next;
        }

        return mergedHead.next;
    }
};
